modules/pubxaiAnalyticsAdapter.js

```diff
@@ -13,196 +23,335 @@ const defaultHost = 'api.pbxai.com';
-      delete data.auctionInit;
-
-      data.pmcDetail = {};
-      Object.assign(data.pmcDetail, {
-        bidDensity: storage ? storage.getItem('pbx:dpbid') : null,
-        maxBid: storage ? storage.getItem('pbx:mxbid') : null,
-        auctionId: storage ? storage.getItem('pbx:aucid') : null,
-      });
+/**
+ * Send auction data back to pubx.ai
+ * @param {string} auctionId the auction to send
+ */
+const send = (auctionId) => {
+  const auctionData = Object.assign({}, auctionCache[auctionId]);
+  if (!shouldFireEventRequest(auctionId, initOptions.samplingRate)) {
+    return;
+  }
+  [
+    {
+      path: winningBidPath,
+      requiredKeys: [
+        'winningBid',
+        'pageDetail',
+        'deviceDetail',
+        'floorDetail',
+        'auctionDetail',
+        'userDetail',
+        'pmacDetail',
+        'initOptions',
+      ],
+      eventType: 'win',
+    },
+    {
+      path: auctionPath,
+      requiredKeys: [
+        'bids',
+        'pageDetail',
+        'deviceDetail',
+        'floorDetail',
+        'auctionDetail',
+        'userDetail',
+        'pmacDetail',
+        'initOptions',
+      ],
+      eventType: 'auction',
+    },
+  ].forEach(({ path, requiredKeys, eventType }) => {
+    const data = Object.fromEntries(
+      requiredKeys.map((key) => [key, auctionData[key]])
+    );
+    if (
+      auctionCache[auctionId].sentAs.includes(eventType) ||
+      !requiredKeys.every((key) => !!auctionData[key])
+    ) {
+      return;
     }
-    data.deviceDetail = {};
-    Object.assign(data.deviceDetail, {
-      platform: navigator.platform,
-      deviceType: getDeviceType(),
-      deviceOS: getOS(),
-      browser: getBrowser()
-    });
-
-    let pubxaiAnalyticsRequestUrl = buildUrl({
+    const pubxaiAnalyticsRequestUrl = buildUrl({
       protocol: 'https',
-      hostname: (initOptions && initOptions.hostName) || defaultHost,
-      pathname: status == 'bidwon' ? winningBidPath : auctionPath,
+      hostname:
+        (auctionData.initOptions && auctionData.initOptions.hostName) ||
+        defaultHost,
+      pathname: path,
       search: {
-        auctionTimestamp: auctionTimestamp,
+        auctionTimestamp: auctionData.auctionDetail.timestamp,
         pubxaiAnalyticsVersion: pubxaiAnalyticsVersion,
-        prebidVersion: getGlobal().version
-      }
+        prebidVersion: getGlobal().version,
+      },
     });
-    if (status == 'bidwon') {
-      ajax(pubxaiAnalyticsRequestUrl, undefined, JSON.stringify(data), { method: 'POST', contentType: 'text/json' });
-    } else if (status == 'auctionEnd' && auctionCache.indexOf(data.initOptions.auctionId) === -1) {
-      ajax(pubxaiAnalyticsRequestUrl, undefined, JSON.stringify(data), { method: 'POST', contentType: 'text/json' });
-      auctionCache.push(data.initOptions.auctionId);
-    }
-  }
-}
+    const payload = new Blob([JSON.stringify(data)], {
+      type: 'text/json',
+    });
+    navigator.sendBeacon(pubxaiAnalyticsRequestUrl, payload);
```

Comment:
instead of sending the data for every auction, we can batch the data and call it once it exceeds certain size.

Reply:
the lambda function which receives this data is not ready to handle batch data, and beacons are not designed for batching, rather they are designed for very small payloads as the browser session "unloads". Additionally, this method of sending data via beacon is consistent with other approaches in pubx:
https://github.com/Pubx-ai/stream/pull/6/files
So, i have made no changes

Reply:
@nathan-pubx beacon supports payload upto 64KB. Even in the existing code base we batch and send logs when visibility change event is triggered(i.e when user leaves page) and not for every auction. lets add batching to this PR right away. In the mean time this PR gets approved in prebid, we will be updating the lambda's to support this. 

```
if (hasBrowserSupport()) {
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "hidden") {
        pushLogs(aucErrorLogs, ERROR_LOG_API);
        pushLogs(pageErrorLogs, ERROR_LOG_API);
        pushLogs(aucEventLogs, EVENT_LOG_API);
      }
    });
  }
```

Reply:
@tej656 (and @pnhegde , since you also asked a question about batching data)

I assume/hope this is in reference to this, more recent, [pull request](https://github.com/Pubx-ai/stream/pull/6/files#diff-6147e3c1761df71fd40952dbcbac388a45f80b8c318f367ef41048351a2c0c99R537), for remote logging in the stream repo?

In which case, i've pushed review actions to implement similar behaviour, plus an additional test to check the batching logic. It's worth noting explicitly (although i'm sure you're aware) that the batch behaviour is not yet supported by the lambda function which is expected to process the payload. Additionally, I haven't changed the payload _at all_, so it can definitely be further optimised for smaller data transfer and marginally reduced cost. Let me know if you'd like me to make additional changes.

Reply:
@nathan-pubx Thanks. We have an update for the lambda to handle the batched data already.